src/Microsoft.AspNetCore.Routing/Patterns/DefaultRoutePatternTransformer.cs

```diff
@@ -0,0 +1,227 @@
+                    // Ex: Admin/{controller=Home}/{action=Index}/{id?} defaults: { area = "Admin" }- with required values: { area = "Blog", ... }
+                    // OR (less likely)
+                    // Ex: Admin/{controller=Home}/{action=Index}/{id?} with required values: { page = "/Index", ... }
+                    return null;
+                }
+            }
+
+            List<RoutePatternParameterPart> updatedParameters = null;
+            List<RoutePatternPathSegment> updatedSegments = null;
+            RouteValueDictionary updatedDefaults = null;
+
+            // So if we get here, we're ready to update the route pattern. We need to update two things:
+            // 1. Remove any default values that conflict with the required values.
+            // 2. Merge any existing required values
+            foreach (var kvp in requiredValues)
+            {
+                var parameter = original.GetParameter(kvp.Key);
+
+                // We only need to handle the case where the required value maps to a parameter. That's the only
+                // case where we allow a default and a required value to disagree, and we already validated the
+                // other cases.
+                if (parameter != null && 
+                    original.Defaults.TryGetValue(kvp.Key, out var defaultValue) && 
+                    !RouteValueEqualityComparer.Default.Equals(kvp.Value, defaultValue))
+                {
+                    if (updatedDefaults == null && updatedSegments == null && updatedParameters == null)
+                    {
+                        updatedDefaults = new RouteValueDictionary(original.Defaults);
+                        updatedSegments = new List<RoutePatternPathSegment>(original.PathSegments);
+                        updatedParameters = new List<RoutePatternParameterPart>(original.Parameters);
+                    }
+
+                    updatedDefaults.Remove(kvp.Key);
+                    RemoveParameterDefault(updatedSegments, updatedParameters, parameter);
+                }
+            }
+
+            foreach (var kvp in original.RequiredValues)
+            {
+                requiredValues.TryAdd(kvp.Key, kvp.Value);
+            }
+
+            return new RoutePattern(
+                original.RawText,
+                updatedDefaults ?? original.Defaults,
+                original.ParameterPolicies,
+                requiredValues,
+                updatedParameters ?? original.Parameters, 
+                updatedSegments ?? original.PathSegments);
+        }
+
+        private bool MatchesConstraints(RoutePattern pattern, RoutePatternParameterPart parameter, string key, RouteValueDictionary requiredValues)
+        {
+            if (pattern.ParameterPolicies.TryGetValue(key, out var policies))
+            {
+                for (var i = 0; i < policies.Count; i++)
+                {
+                    var policy = _policyFactory.Create(parameter, policies[i]);
```

Comment:
I think route constraints today were created twice - once to denormalize routes and again by the matcher. I think there is an opportunity to cache the created route constraints and reuse them now that both sides of the code are in routing.

Low priority tho

Reply:
I super don't want to do this for layering reasons.